File: rust/src/bin/day12.rs
use std::{
    collections::{BTreeMap, BTreeSet, VecDeque},
    io::BufRead,
};

use anyhow::Result;

use rust::read;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
enum Edge {
    Vertical(usize, usize),
    Horizontal(usize, usize),
}

#[derive(Debug)]
struct Map {
    h: usize,
    w: usize,
    data: Vec<Vec<char>>,
}

fn adjacent_edges((x, y): (usize, usize)) -> [Edge; 4] {
    [
        Edge::Horizontal(x, y),
        Edge::Horizontal(x + 1, y),
        Edge::Vertical(x, y),
        Edge::Vertical(x, y + 1),
    ]
}

impl Map {
    fn good_point(&self, (x, y): &(i32, i32)) -> bool {
        *x >= 0 && *x < self.h as i32 && *y >= 0 && *y < self.w as i32
    }

    fn adjacent_points(&self, (x, y): (usize, usize)) -> Vec<(usize, usize)> {
        let x = x as i32;
        let y = y as i32;
        [(x, y - 1), (x, y + 1), (x - 1, y), (x + 1, y)]
            .into_iter()
            .filter(|pos| self.good_point(pos))
            .map(|(x, y)| (x as usize, y as usize))
            .collect()
    }

    fn is_separated_edge(&self, edge: &Edge) -> bool {
        let (lhs, rhs) = match edge {
            Edge::Vertical(x, y) => {
                let x = *x as i32;
                let y = *y as i32;

                let lhs = (x, y - 1);
                let rhs = (x, y);

                (lhs, rhs)
            }
            Edge::Horizontal(x, y) => {
                let x = *x as i32;
                let y = *y as i32;

                let lhs = (x - 1, y);
                let rhs = (x, y);

                (lhs, rhs)
            }
        };

        if self.good_point(&lhs) && self.good_point(&rhs) {
            self.data[lhs.0 as usize][lhs.1 as usize] != self.data[rhs.0 as usize][rhs.1 as usize]
        } else {
            true
        }
    }

    fn visit(
        &self,
        pos: (usize, usize),
        colors: &mut [Vec<Option<i32>>],
    ) -> (BTreeSet<(usize, usize)>, BTreeSet<Edge>) {
        let datum = self.data[pos.0][pos.1];
        let color = colors[pos.0][pos.1];

        let mut visited = BTreeSet::new();
        let mut edges = BTreeSet::new();

        let mut queue = VecDeque::from([pos]);
        while let Some(pos) = queue.pop_front() {
            if visited.contains(&pos) {
                continue;
            }
            visited.insert(pos);
            edges.extend(
                adjacent_edges(pos)
                    .into_iter()
                    .filter(|e| self.is_separated_edge(e)),
            );

            let adjacents: Vec<(usize, usize)> = self
                .adjacent_points(pos)
                .into_iter()
                .filter(|(x, y)| self.data[*x][*y] == datum)
                .collect();
            for (x, y) in &adjacents {
                colors[*x][*y] = color;
            }
            queue.extend(adjacents);
        }

        (visited, edges)
    }

    fn price(&self, pos: (usize, usize), colors: &mut [Vec<Option<i32>>]) -> usize {
        let (visited, edges) = self.visit(pos, colors);
        let area = visited.len();
        let perimeter = edges
            .into_iter()
            .map(|edge| if self.is_separated_edge(&edge) { 1 } else { 0 })
            .sum::<usize>();

        area * perimeter
    }

    fn prices(self) -> usize {
        let mut colors: Vec<Vec<Option<i32>>> =
            vec![vec![None; self.data[0].len()]; self.data.len()];

        let mut ans = 0;
        let mut color = 0;
        for x in 0..self.data.len() {
            for y in 0..self.data[0].len() {
                if colors[x][y].is_some() {
                    continue;
                }
                colors[x][y] = Some(color);
                let price = self.price((x, y), &mut colors);
                ans += price;
                color += 1;
            }
        }

        ans
    }

    fn edge_same_side(&self, lhs: &Edge, rhs: &Edge) -> bool {
        match (lhs.clone(), rhs.clone()) {
            (Edge::Vertical(x1, y1), Edge::Vertical(x2, y2))
                if y1 == y2 && (x1 == x2 + 1 || x2 == x1 + 1) =>
            {
                (y1 > 0 && self.data[x1][y1 - 1] == self.data[x2][y2 - 1])
                    || (y1 < self.w && self.data[x1][y1] == self.data[x2][y2])
            }
            (Edge::Horizontal(x1, y1), Edge::Horizontal(x2, y2))
                if x1 == x2 && (y1 == y2 + 1 || y2 == y1 + 1) =>
            {
                (x1 > 0 && self.data[x1 - 1][y1] == self.data[x2 - 1][y2])
                    || (x1 < self.h && self.data[x1][y1] == self.data[x2][y2])
            }
            _ => false,
        }
    }

    fn price2(&self, pos: (usize, usize), colors: &mut [Vec<Option<i32>>]) -> usize {
        let (visited, edges) = self.visit(pos, colors);
        let area = visited.len();

        let mut groups: BTreeMap<usize, Vec<Edge>> = BTreeMap::new();
        let mut sides = 0;
        for edge in edges.into_iter() {
            if let Some((_, g)) = groups
                .iter_mut()
                .find(|(_, g)| g.iter().any(|e| self.edge_same_side(&edge, e)))
            {
                g.push(edge)
            } else {
                groups.insert(sides, vec![edge]);
                sides += 1;
            }
        }

        area * sides
    }

    fn prices2(self) -> usize {
        let mut colors: Vec<Vec<Option<i32>>> =
            vec![vec![None; self.data[0].len()]; self.data.len()];

        let mut ans = 0;
        let mut color = 0;
        for x in 0..self.data.len() {
            for y in 0..self.data[0].len() {
                if colors[x][y].is_some() {
                    continue;
                }
                colors[x][y] = Some(color);
                let price = self.price2((x, y), &mut colors);
                ans += price;
                color += 1;
            }
        }

        ans
    }
}

fn read_map() -> Result<Map> {
    let reader = read(12, false)?;
    let data: Vec<Vec<char>> = reader
        .lines()
        .map_while(Result::ok)
        .filter(|line| !line.is_empty())
        .map(|line| line.chars().collect())
        .collect();
    Ok(Map {
        h: data.len(),
        w: data[0].len(),
        data,
    })
}

fn part1() -> Result<()> {
    let map = read_map()?;
    println!("{}", map.prices());
    Ok(())
}

fn part2() -> Result<()> {
    let map = read_map()?;
    println!("{}", map.prices2());
    Ok(())
}

fn main() -> Result<()> {
    part1()?;
    part2()?;
    Ok(())
}

File: python/day12.py
import dataclasses
import enum
import itertools
from functools import cached_property

import utils


class Side(enum.IntEnum):
    Horizontal = 0
    Vertical = 1


@dataclasses.dataclass(order=True)
class Edge:
    side: Side
    x: int
    y: int


@dataclasses.dataclass(unsafe_hash=True)
class Point:
    x: int
    y: int

    def shared_edge(self, point: "Point") -> Edge | None:
        if self.x == point.x:
            if self.y == point.y + 1:
                return Edge(Side.Vertical, self.x, self.y)
            if point.y == self.y + 1:
                return Edge(Side.Vertical, self.x, point.y)

        if self.y == point.y:
            if self.x == point.x + 1:
                return Edge(Side.Horizontal, self.x, self.y)
            if point.x == self.x + 1:
                return Edge(Side.Horizontal, point.x, self.y)

        return None


@dataclasses.dataclass
class Map:
    data: list[str]

    @cached_property
    def h(self) -> int:
        return len(self.data)

    @cached_property
    def w(self) -> int:
        return len(self.data[0])

    def inside(self, p: Point) -> bool:
        return p.x >= 0 and p.x < self.h and p.y >= 0 and p.y < self.w

    def adjacents(self, point: Point) -> list[Point]:
        return [
            Point(point.x, point.y - 1),
            Point(point.x, point.y + 1),
            Point(point.x - 1, point.y),
            Point(point.x + 1, point.y),
        ]

    def same(self, lhs: Point, rhs: Point) -> bool:
        if not self.inside(lhs) or not self.inside(rhs):
            return False
        return self.data[lhs.x][lhs.y] == self.data[rhs.x][rhs.y]

    def connectable(self, lhs: Edge, rhs: Edge) -> bool:
        if lhs.side != rhs.side:
            return False

        match lhs.side:
            case Side.Horizontal if lhs.x == rhs.x and abs(lhs.y - rhs.y) <= 1:
                p1, p2 = Point(lhs.x, lhs.y), Point(rhs.x, rhs.y)
                p3, p4 = Point(lhs.x - 1, lhs.y), Point(rhs.x - 1, rhs.y)

                return (
                    (not self.inside(p1) and not self.inside(p2) and self.same(p3, p4))
                    or (
                        not self.inside(p3)
                        and not self.inside(p4)
                        and self.same(p1, p2)
                    )
                    or (self.same(p1, p2) or self.same(p3, p4))
                )

            case Side.Vertical if lhs.y == rhs.y and abs(lhs.x - rhs.x) <= 1:
                p1, p2 = Point(lhs.x, lhs.y), Point(rhs.x, rhs.y)
                p3, p4 = Point(lhs.x, lhs.y - 1), Point(rhs.x, rhs.y - 1)

                return (
                    (not self.inside(p1) and not self.inside(p2) and self.same(p3, p4))
                    or (
                        not self.inside(p3)
                        and not self.inside(p4)
                        and self.same(p1, p2)
                    )
                    or (self.same(p1, p2) or self.same(p3, p4))
                )

        return False

    def visit(self, point: Point) -> set[Point]:
        visited: set[Point] = set()
        visited.add(point)

        def dfs(p: Point) -> None:
            for adj in filter(self.inside, self.adjacents(p)):
                if adj in visited or not self.same(point, adj):
                    continue
                visited.add(adj)
                dfs(adj)

        dfs(point)

        return visited


def read_map() -> Map:
    with utils.read(day=12, sample=False) as f:
        data = [line.strip() for line in f.readlines() if len(line.strip()) > 0]
    return Map(data)


def part1() -> None:
    m = read_map()
    visited: set[Point] = set()

    ans = 0

    for x, y in itertools.product(range(m.h), range(m.w)):
        point = Point(x, y)
        if point in visited:
            continue

        local_visited = m.visit(point)

        area = len(local_visited)
        perimeter = 0

        for p in local_visited:
            for adj in m.adjacents(p):
                if not m.same(p, adj):
                    perimeter += 1

        ans += area * perimeter

        visited = visited | local_visited

    print(ans)


def part2() -> None:
    m = read_map()
    visited: set[Point] = set()

    ans = 0

    for x, y in itertools.product(range(m.h), range(m.w)):
        point = Point(x, y)
        if point in visited:
            continue

        local_visited = m.visit(point)

        area = len(local_visited)

        edges: list[Edge] = []

        for p in local_visited:
            for adj in m.adjacents(p):
                if not m.same(p, adj):
                    edge = p.shared_edge(adj)
                    assert edge is not None
                    edges.append(edge)

        edges = sorted(edges)

        connected_edges: list[list[Edge]] = []

        for e in edges:
            found = False
            for cedges in connected_edges:
                if any(m.connectable(e, ce) for ce in cedges):
                    found = True
                    cedges.append(e)
            if not found:
                connected_edges.append([e])

        ans += area * len(connected_edges)

        visited = visited | local_visited

    print(ans)


if __name__ == "__main__":
    part1()
    part2()
